core/src/main/resources/xsl/raxRoles.xsl

```diff
@@ -113,15 +113,24 @@
 
     <xsl:template name="generateRoles">
         <xsl:param name="roles" as="xsd:string*" select="()"/>
-        <xsl:if test="not('#all' = $roles)">
-            <xsl:for-each select="$roles">
-                <wadl:param name="X-ROLES" style="header" rax:code="403" rax:anyMatch="true"
-                            rax:message="You are forbidden to perform the operation" type="xsd:string"
-                            required="true" repeating="true">
-                    <xsl:attribute name="fixed" select="rax:transformNBSP(.)"/>
-                </wadl:param>
-            </xsl:for-each>
-        </xsl:if>
+        <xsl:choose>
+            <xsl:when test="not('#all' = $roles)">
+                <rax:captureHeader name="X-RELEVANT-ROLES">
+                    <xsl:attribute name="path" select="string-join(('for $h in req:headers(''X-ROLES'', true()) return if ($h = (''', string-join($roles, ''', '''), ''')) then $h else ()'), '')"/>
```

Comment:
I'd put this into different steps using variables.  For example the first thing you want  to do is put quotes on all the roles right?

````xml
  <xsl:variable name="quoted" as="xsd:string*" select="for $r in $roles return concat(''',$r,''')"/>
````
You actually missed calling ```rax:transformNBSP``` for each role, because you want to list all non-breaking spaces into spaces.  So it may be better to say...

````xml
  <xsl:variable name="quoted" as="xsd:string*" select="for $r in $roles return concat(''',rax:transformNBSP($r),''')"/>
````
 You then want to form the sequence sperated by spaces...

````xml
<xsl:variable name="sequence" as="xsd:string"><xsl:value-of select="$quoted" separator=", "/></xsl:variable>
````
This should give you something like 'role1', 'role2', 'role3'.

The other neat trick is that you can avoid using ```xsl:attribute``` by using curly braces in an XML attribute of a template.  This does interpolation of variables or even full expressions. That also avoids things like string-joins and concats..

````xml
    <rax:captureHeader name="X-RELEVANT-ROLES" path="for $h in req:headers('X-ROLES', true) return if ($h = ({$sequence}) then $h else ()" />
````